Add a first version of new upgrade docs (#4750)
digital-asset/daml: docs/source/upgrade/index.rst

archive the old version of a contract and create a new contract
instead.

Structuring upgrade contracts
=============================

Upgrade contracts are specific to the templates that are being
upgraded. However, there are common patterns between most of them. We
use the example of a simple *Coin* template as an example here.

.. literalinclude:: daml/Upgrade.daml
  :language: daml
  :start-after: -- COIN_BEGIN
  :end-before: -- COIN_END

A *Coin* has an issuer and an owner and both are signatories.  Our
goal is to extend this *Coin* template with a field that represents
the number of coins to avoid needing 1000 contracts to represent 1000
coins. (In a real application, you would also want choices for merging
and splitting such a *Coin*. For presentational purposes, we omit
those here.) We use a different name for the new template here. This
is not required as templates are identified by the triple
``(PackageId, ModuleName, TemplateName)`` but here we omit package ids
and module names for presentational purposes.

.. literalinclude:: daml/Upgrade.daml
  :language: daml
  :start-after: -- COIN_AMOUNT_BEGIN
  :end-before: -- COIN_AMOUNT_END

> Now, that we defined our new template, we need to provide a way for

Reviewer: ```suggestion
Next, we need to provide a way for
```

> the signatories issuer and owner to agree to a contract being

Reviewer: ```suggestion
the signatories, issuer and owner, to agree to a contract being
```

upgraded. It would be possible to structure this such that issuer and

> owner have to agree to an upgrade for each *Coin* contract

Reviewer: ```suggestion
owner have to agree to an upgrade for each individual *Coin* contract
```

> separately. However, since the template definition for those is all

Reviewer: ```suggestion
separately. However, since the template definition for all of them is
```

the same, this is usually not necessary for most
applications. Instead, we collect agreement from the signatories only
once and use that to upgrade all coins. Since there are multiple
signatories involved here, we use a :ref:`Propose-Accept workflow
<intro propose accept>`. First, we define an *UpgradeCoinProposal*
template that will be created by the issuer. This template has an
*Accept* choice that the *owner* can exercise which will then create
an *UpgradeCoinAgreement*.

.. literalinclude:: daml/Upgrade.daml
  :language: daml
  :start-after: -- UPGRADE_PROPOSAL_BEGIN
  :end-before: -- UPGRADE_PROPOSAL_END

Now we can define the *UpgradeCoinAgreement* template. This template

> has one *nonconsuming* choice that accepts a *Coin* contract, archives

Reviewer: The word "accept" is a bit overloaded here. We also need to make it clear that we pass the contract by reference not by value.
```suggestion
has one *nonconsuming* choice that takes the contract ID of a *Coin* contract, archives
```

this *Coin* contract and creates a *CoinWithAmount* contract with

> *amount* set to 1.

Reviewer: ```suggestion
the same issuer and owner and the *amount* set to 1.
```
